fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+      flux *= 0.5;
+   }
+};
+
+// Upwind Flux, Not Yet Implemented
+class UpwindFlux : public RiemannSolver
+{
+public:
+   // Upwind Flux, Not Yet Implemented
+   void Eval(const Vector &state1, const Vector &state2, const Vector &fluxN1,
+             const Vector &fluxN2, const double maxE, const Vector &nor,
+             Vector &flux)
+   {
+      // NOTE: nor in general is not a unit normal
+      mfem_error("Not Implemented");
+   }
+};
+
+class AdvectionFormIntegrator : public HyperbolicFormIntegrator
+{
+private:
+   VectorCoefficient &b;  // velocity coefficient
+   Vector bval;           // velocity value storage
+
+public:
+   /**
+    * @brief Compute F(u)
+    *
+    * @param U U (u) at current integration point
+    * @param Tr current element transformation with integration point
+    * @param FU F(u) = ubᵀ
+    * @return double maximum characteristic speed, |b|
+    */
+   double ComputeFlux(const Vector &U, ElementTransformation &Tr,
+                      DenseMatrix &FU)
+   {
+      b.Eval(bval, Tr, Tr.GetIntPoint());
+      MultVWt(U, bval, FU);
+      return bval.Norml2();
+   }
+   /**
+    * @brief Compute normal flux, F(u)n
+    *
+    * @param U U (u) at current integration point
+    * @param normal normal vector, usually not a unit vector
+    * @param Tr current element transformation with integration point
+    * @param FUdotN F(u)n = u(b⋅n)
+    * @return double maximum characteristic speed, |b|
+    */
+   double ComputeFluxDotN(const Vector &U, const Vector &normal,
+                          ElementTransformation &Tr, Vector &FUdotN)
+   {
+      b.Eval(bval, Tr, Tr.GetIntPoint());
+      const double bN = bval * normal;
+      FUdotN = U;
+      FUdotN *= bN;
+      return bval.Norml2();
+   }
+
+   /**
+    * @brief Construct a new Advection Element Form Integrator object with given
+    * integral order offset
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param dim spatial dimension
+    * @param b_ velocity coefficient, possibly depends on space
+    * @param IntOrderOffset_ 2*p + IntOrderOffset will be used for quadrature
+    */
+   AdvectionFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                           VectorCoefficient &b_,
+                           const int IntOrderOffset_ = 3)
+      : HyperbolicFormIntegrator(rsolver_, dim, 1, IntOrderOffset_), b(b_),
+        bval(dim) {}
+   /**
+    * @brief Construct a new Advection Element Form Integrator object with given
+    * integral rule
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param dim spatial dimension
+    * @param b_ velocity coefficient, possibly depends on space
+    * @param ir this integral rule will be used for the Gauss quadrature
+    */
+   AdvectionFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                           VectorCoefficient &b_,
+                           const IntegrationRule *ir)
+      : HyperbolicFormIntegrator(rsolver_, dim, 1, ir), b(b_), bval(dim) {}
```

Comment:
I would suggest moving the constructors up to the top of the class so they are easier to find. Also, I'm not sure we need both the `IntOrderOffset` and `ir` versions of all of the constructors.